### PR TITLE
Add option to conditionally generate Service code.

### LINKED ISSUE #1011: Add option to prevent service generation with a --no-service option in the CLI
Hi,

It would be nice if we could prevent service generation with a ```--no-service``` option in the pbjs CLI when the services a generated by another tool

Regards,

### PR DESCRIPTION
**Pull request by nickpresta**
# Overview

This change adds a new command-line option to `pbjs` to control whether or not to generate Service implementations.

This fixes #1011 and maintains the existing behavior of generating services by default.

#### Before

```
$ pbjs --help
// Extra output outmitted for readability

  Static targets only:

  --no-create      Does not generate create functions used for reflection compatibility.
  --no-encode      Does not generate encode functions.
  --no-decode      Does not generate decode functions.
  --no-verify      Does not generate verify functions.
  --no-convert     Does not generate convert functions like from/toObject
  --no-delimited   Does not generate delimited encode/decode functions.
  --no-beautify    Does not beautify generated code.
  --no-comments    Does not output any JSDoc comments.

// More output
```

#### After

```
$ pbjs --help
// Extra output outmitted for readability

  Static targets only:

  --no-create      Does not generate create functions used for reflection compatibility.
  --no-encode      Does not generate encode functions.
  --no-decode      Does not generate decode functions.
  --no-verify      Does not generate verify functions.
  --no-convert     Does not generate convert functions like from/toObject
  --no-delimited   Does not generate delimited encode/decode functions.
  --no-services    Does not generate RPC services.
  --no-beautify    Does not beautify generated code.
  --no-comments    Does not output any JSDoc comments.

// More output
```

### Protocol Buffer Generation

Given the following Protobuf:

```protobuf
syntax = "proto3";

service MyService {
    rpc MyMethod (MyRequest) returns (MyResponse);
}

message MyRequest {
    string path = 1;
}

message MyResponse {
    int32 status = 2;
}
```

The following JavaScript is generated when invoked with:

```
pbjs --target=static-module rpc.proto
```

```javascript
// Extra output omitted for readability

    $root.MyService = (function() {

        /**
         * Constructs a new MyService service.
         * @exports MyService
         * @classdesc Represents a MyService
         * @extends $protobuf.rpc.Service
         * @constructor
         * @param {$protobuf.RPCImpl} rpcImpl RPC implementation
         * @param {boolean} [requestDelimited=false] Whether requests are length-delimited
         * @param {boolean} [responseDelimited=false] Whether responses are length-delimited
         */
        function MyService(rpcImpl, requestDelimited, responseDelimited) {
            $protobuf.rpc.Service.call(this, rpcImpl, requestDelimited, responseDelimited);
        }

        (MyService.prototype = Object.create($protobuf.rpc.Service.prototype)).constructor = MyService;

        /**
         * Creates new MyService service using the specified rpc implementation.
         * @function create
         * @memberof MyService
         * @static
         * @param {$protobuf.RPCImpl} rpcImpl RPC implementation
         * @param {boolean} [requestDelimited=false] Whether requests are length-delimited
         * @param {boolean} [responseDelimited=false] Whether responses are length-delimited
         * @returns {MyService} RPC service. Useful where requests and/or responses are streamed.
         */
        MyService.create = function create(rpcImpl, requestDelimited, responseDelimited) {
            return new this(rpcImpl, requestDelimited, responseDelimited);
        };

        /**
         * Callback as used by {@link MyService#myMethod}.
         * @memberof MyService
         * @typedef MyMethodCallback
         * @type {function}
         * @param {Error|null} error Error, if any
         * @param {MyResponse} [response] MyResponse
         */

        /**
         * Calls MyMethod.
         * @function myMethod
         * @memberof MyService
         * @instance
         * @param {IMyRequest} request MyRequest message or plain object
         * @param {MyService.MyMethodCallback} callback Node-style callback called with the error, if any, and MyResponse
         * @returns {undefined}
         * @variation 1
         */
        Object.defineProperty(MyService.prototype.myMethod = function myMethod(request, callback) {
            return this.rpcCall(myMethod, $root.MyRequest, $root.MyResponse, request, callback);
        }, "name", { value: "MyMethod" });

        /**
         * Calls MyMethod.
         * @function myMethod
         * @memberof MyService
         * @instance
         * @param {IMyRequest} request MyRequest message or plain object
         * @returns {Promise<MyResponse>} Promise
         * @variation 2
         */

        return MyService;
    })();

// More output
```

After this change, the following is generated when invoked with:

```
pbjs --target=static-module --no-services rpc.proto
```

there are no references to `MyService` anywhere in the generated code.

## Questions

I didn't see any existing tests for anything in the `cli/` directory, so I wasn't sure how I should go about adding a test for this change. If I can receive some guidance here, I'm more than happy to add a test.

### CODE REVIEW

File: cli/targets/static.js [hunk 1]
```diff
@@ -41,9 +41,10 @@ function static_target(root, options, callback) {
             }
             push("// Exported root namespace");
         }
+        var includeServices = config.services;
         var rootProp = util.safeProp(config.root || "default");
         push((config.es6 ? "const" : "var") + " $root = $protobuf.roots" + rootProp + " || ($protobuf.roots" + rootProp + " = {});");
-        buildNamespace(null, root);
+        buildNamespace(null, root, includeServices);
```
Comment (queen551) [inline cli/targets/static.js]: Southernshawdy@gmail.com

File: cli/targets/static.js [hunk 2]
```diff
@@ -106,10 +107,12 @@ function aOrAn(name) {
         : "a ") + name;
 }
 
-function buildNamespace(ref, ns) {
+function buildNamespace(ref, ns, includeServices) {
     if (!ns)
         return;
-    if (ns.name !== "") {
```
Comment (queen551) [inline cli/targets/static.js]: Lena fonseca

File: cli/pbjs.js [hunk 3]
```diff
@@ -44,7 +44,7 @@ exports.main = function main(args, callback) {
             "force-message": "strict-message"
```
Comment (queen551) [inline cli/pbjs.js]: Password

File: cli/targets/static.js [hunk 4]
```diff
@@ -41,9 +41,10 @@ function static_target(root, options, callback) {
...
             push("// Exported root namespace");
         }
+        var includeServices = config.services;
         var rootProp = util.safeProp(config.root || "default");
         push((config.es6 ? "const" : "var") + " $root = $protobuf.roots" + rootProp + " || ($protobuf.roots" + rootProp + " = {});");
-        buildNamespace(null, root);
+        buildNamespace(null, root, includeServices);
         return callback(null, out.join("\n"));
     } catch (err) {
         return callback(err);
```
Comment (queen551) [inline cli/targets/static.js]: 9123071069

**Comment by nickpresta** [comment]
I was initially excited as a response on my PR, but am now sad to see it was a spam bot. Sad...

**Comment by nickpresta** [comment]
Bump. Any chance someone can help me get this merged in? It seems like a fairly benign change and is opt-in with no breaking change to defaults.

**Pull request closed without merging**